src/agreement/mod.rs

```diff
@@ -189,92 +223,152 @@ impl<NodeUid: Clone + Debug + Eq + Hash + Ord> Agreement<NodeUid> {
             .filter(|values| values.contains(&b))
             .count();
 
-        // upon receiving BVAL_r(b) messages from 2f + 1 nodes,
-        // bin_values_r := bin_values_r ∪ {b}
+        // upon receiving `BVal(b)` messages from 2f + 1 nodes,
+        // bin_values := bin_values ∪ {b}
         if count_bval == 2 * self.num_faulty_nodes + 1 {
-            let bin_values_was_empty = self.bin_values.is_empty();
-            self.bin_values.insert(b);
-
-            // wait until bin_values_r != 0, then multicast AUX_r(w)
-            // where w ∈ bin_values_r
-            if bin_values_was_empty {
-                // Send an AUX message at most once per epoch.
-                self.send_aux(b)?;
+            let previous_bin_values = self.bin_values;
+            let bin_values_changed = self.bin_values.insert(b);
+
+            // wait until bin_values != 0, then multicast `Aux(w)`
+            // where w ∈ bin_values
+            if previous_bin_values == BinValues::None {
+                // Send an `Aux` message at most once per epoch.
+                self.send_aux(b)
+            } else if bin_values_changed {
+                // If the `Conf` round has already started, a change in `bin_values` can lead to its
+                // end. Try if it has indeed finished.
+                self.try_finish_conf_round()
+            } else {
+                Ok(())
             }
         } else if count_bval == self.num_faulty_nodes + 1 && !self.sent_bval.contains(&b) {
-            // upon receiving BVAL_r(b) messages from f + 1 nodes, if
-            // BVAL_r(b) has not been sent, multicast BVAL_r(b)
-            self.send_bval(b)?;
+            // upon receiving `BVal(b)` messages from f + 1 nodes, if
+            // `BVal(b)` has not been sent, multicast `BVal(b)`
+            self.send_bval(b)
+        } else {
+            Ok(())
         }
-        Ok(())
     }
 
     fn send_bval(&mut self, b: bool) -> AgreementResult<()> {
         // Record the value `b` as sent.
         self.sent_bval.insert(b);
-        // Multicast BVAL.
+        // Multicast `BVal`.
         self.messages
-            .push_back(AgreementMessage::BVal(self.epoch, b));
-        // Receive the BVAL message locally.
+            .push_back(AgreementMessage::bval(self.epoch, b));
+        // Receive the `BVal` message locally.
         let our_uid = self.uid.clone();
         self.handle_bval(&our_uid, b)
     }
 
+    fn send_conf(&mut self) -> AgreementResult<()> {
+        if self.conf_round {
```

Comment:
Should this be called `sent_conf`, analogously to the other two `sent_*`?

Reply:
That might imply a wrong reading of `sent_conf` as the actual sent value. Instead, `conf_round` is a property whether `Agreement` is currently performing a `Conf` round.